### PR TITLE
DM-45860: Do not do the collections query again unless we have to

### PR DESCRIPTION
**Pull request by timj**
@andy-slac does this help?

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

### CODE REVIEW

**Comment by timj** [comment]
@andy-slac I also realized that I had forgotten to add negative limit support to the dimension-records and data-ids command-lines so I've made that change here. Did the collections fix help you?

File: python/lsst/daf/butler/utils.py [hunk 1]
```diff
@@ -134,6 +134,33 @@ def globToRegex(expressions: str | EllipsisType | None | list[str]) -> list[str
     return results
 
 
+def has_globs(expressions: str | EllipsisType | None | Iterable[str]) -> bool:
+    """Determine if the expressions have any glob wildcard characters.
+
+    If a stand-alone '``*``' is found in ``expressions``, or expressions is
+    empty or `None`, then the special value ``...`` will be returned,
+    indicating that any string will match.
```
Comment (andy-slac) [inline python/lsst/daf/butler/utils.py]: This looks like it's copied from `globToRegex` but it probably does not apply here?

File: python/lsst/daf/butler/utils.py [hunk 2]
```diff
@@ -134,6 +134,33 @@ def globToRegex(expressions: str | EllipsisType | None | list[str]) -> list[str
...
+
+    Parameters
+    ----------
+    expressions : `str` or `list` [`str`]
```
Comment (andy-slac) [inline python/lsst/daf/butler/utils.py]: `Iterable` instead of `list`?

**Review by andy-slac — APPROVED** [review]
Looks great.

**Comment by andy-slac** [comment]
> Did the collections fix help you?

Not really, there is a deeper issue when you pass a list of all collection names (mixing all chained and non-chained) to `resolve_wildcard` method, it blows up by flattening all chains again.

**Pull request merged**